python-packages/smithy-core/smithy_core/documents.py

```diff
@@ -284,3 +310,137 @@ def __eq__(self, other: object) -> bool:
         # JsonDocument, cares about the schema then it's responsible for transforming
         # the output of `as_value` to the canonical form.
         return self.as_value() == other.as_value()
+
+
+class DocumentSerializer(ShapeSerializer):
+    """Serializes shapes into document representations."""
+
+    result: Document | None = None
+
+    def expect_result(self) -> Document:
+        """Expect a document to have been serialized and return it."""
+        if self.result is None:
+            raise ExpectationNotMetException(
+                "Expected document serializer to have a result, but was None"
+            )
+        return self.result
+
+    @override
+    @contextmanager
+    def begin_struct(self, schema: "Schema") -> Iterator[ShapeSerializer]:
+        delegate = _DocumentStructSerializer(schema)
+        try:
+            yield delegate
+        finally:
+            self.result = delegate.result
```

Comment:
Why set the result when an error occurs?

Reply:
In this case I don't think it actually matters, but i switched it to try/except/else as well.